common/k8s/src/event_source.rs

```diff
@@ -261,198 +261,239 @@ impl K8sEventStream {
-                                                Ok(watcher::Event::<Pod>::Restarted(_))
-                                            );
-                                            async move { matched }
-                                        })
-                                        .map({
-                                            move |e| {
-                                                match e {
-                                                    Ok(watcher::Event::Deleted(e)) => {
-                                                        info!("previous k8s event logger deleted");
-                                                        // We've detected the old one is dead, now what?
-                                                        delete_time.store(
-                                                            e.metadata
-                                                                .deletion_timestamp
-                                                                .map(|t| {
-                                                                    info!("Ignoring k8s events before {}", t.0 - chrono::Duration::seconds(2));
-                                                                    NonZeroI64::new(t.0.timestamp() - 2)
-                                                                })
-                                                                .flatten(),
+    fn waiter_stream<T>(
+        pod_name: impl Into<String>,
+        namespace: impl Into<String>,
+        client: Arc<Client>,
+        delete_time: Arc<AtomicCell<Option<NonZeroI64>>>,
+    ) -> impl Stream<Item = Result<StreamElem<T>, K8sEventStreamError>> {
+        let pod_name = pod_name.into();
+        let namespace = namespace.into();
+
+        let waiter = move |_| {
+            let pod_name = pod_name.clone();
+            let namespace = namespace.clone();
+            let client = client.clone();
+
+            let delete_time = delete_time.clone();
+            let pods: Api<Pod> = Api::namespaced(client.as_ref().clone(), &namespace);
+            // subscribe to pod api and filter pods by POD_APP_LABEL
+            let pod_label =
+                std::env::var("POD_APP_LABEL").unwrap_or_else(|_| "logdna-agent".into());
+
+            async move {
+                // Find the oldest pod of the latest generation
+                enum Cont {
+                    Cont,
+                    Break,
+                }
+
+                if let Some(oldest_pod_name) =
+                    K8sEventStream::get_oldest_pod(pods.clone(), &pod_label).await?
+                {
+                    if oldest_pod_name == pod_name {
+                        info!("begin logging k8s events");
+                        Ok(None)
+                    } else {
+                        info!("watching {}", oldest_pod_name);
+                        let params = ListParams::default()
+                            .timeout(30)
```

Comment:
After the timer elapses, shouldn't we call watch again / restart the stream?

Reply:
This is configuration for the kube-rs watcher object, it uses this to configured how it handles timeouts on the kubernetes REST api calls